### PR TITLE
Implmentation of warning headers

### PR DESCRIPTION
**Pull request by crobby**
Pass through warning headers which k8s returns. 

Issue:  https://github.com/rancher/rancher/issues/39413

### CODE REVIEW

**Comment by crobby** [comment]
Requires https://github.com/rancher/apiserver/pull/19 to be merged and released first

**Comment by crobby** [comment]
This seems to be working well.  Sample output complements of curl...
```
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!
* We are completely uploaded and fine
< HTTP/2 201 
< content-encoding: gzip
< content-type: application/json
< expires: Wed 24 Feb 1982 18:42:00 GMT
< warning: 299 - would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "nginx" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "nginx" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "nginx" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "nginx" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
< x-api-schemas: https://localhost:9553/v1/schemas
< content-length: 1098
< date: Tue, 20 Dec 2022 19:56:18 GMT

```

File: pkg/client/factory.go [hunk 1]
```diff
@@ -111,52 +111,70 @@ func (p *Factory) AdminK8sInterface() (kubernetes.Interface, error) {
...
+	config := *p.clientCfg
+	config.WarningHandler = warningHandler
+	return newDynamicClient(ctx, &config, p.impersonate)
 }
 
-func (p *Factory) Client(ctx *types.APIRequest, s *types.APISchema, namespace string) (dynamic.ResourceInterface, error) {
-	return newClient(ctx, p.clientCfg, s, namespace, p.impersonate)
+func (p *Factory) Client(ctx *types.APIRequest, s *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error) {
+	config := *p.clientCfg
+	config.WarningHandler = warningHandler
```
Comment (cmurphy) [inline pkg/client/factory.go:122]: Could this just be added to `newClient` and/or `newDynamicClient`?

Comment (crobby) [reply pkg/client/factory.go:122]: > Could this just be added to `newClient` and/or `newDynamicClient`?

That might be a slightly cleaner way to go.

Comment (crobby) [reply pkg/client/factory.go]: Ok, I switched it up a bit and added it to newDynamicClient.  I think I like that a bit better.

**Comment by crobby** [comment]
Looks like a little rebasing is also in order.  I'll work on that soon.

File: pkg/stores/partition/store.go [hunk 2]
```diff
@@ -384,7 +385,7 @@ func toAPIEvent(apiOp *types.APIRequest, schema *types.APISchema, event watch.Ev
 		return apiEvent
 	}
 
-	apiEvent.Object = toAPI(schema, event.Object)
+	apiEvent.Object = toAPI(schema, event.Object, nil)
```
Comment (cmurphy) [inline pkg/stores/partition/store.go:388]: In the first version of this, the `toAPIEvent` function accepted warnings, are they not needed any more for events?

Comment (crobby) [reply pkg/stores/partition/store.go:388]: I don't think it's needed/useful in our case.  I'll keep exploring though.  It might be back.

**Comment by crobby** [comment]
Updated to reflect new apiserver version after that change merged

File: go.mod [hunk 3]
```diff
@@ -11,103 +11,103 @@ replace (
...
-	k8s.io/api v0.25.4
-	k8s.io/apiextensions-apiserver v0.25.4
-	k8s.io/apimachinery v0.25.4
-	k8s.io/apiserver v0.25.4
+	github.com/sirupsen/logrus v1.9.0
+	github.com/stretchr/testify v1.8.1
+	github.com/urfave/cli v1.22.10
+	github.com/urfave/cli/v2 v2.23.7
+	golang.org/x/sync v0.1.0
+	k8s.io/api v0.26.0
```
Comment (maxsokolovsky) [inline go.mod:32]: Are we ready to use the 0.26 versions of the K8s modules?

Comment (crobby) [reply go.mod:32]: Hmm, good question.  It slipped in with my mod updating.  I can pull it back for now if we're not sure.

Comment (crobby) [reply go.mod]: Changed to only bump rancher/apiserver

**Review by cmurphy — APPROVED** [review]

**Review by maxsokolovsky — CHANGES_REQUESTED** [review]

File: pkg/stores/proxy/proxy_store.go [hunk 4]
```diff
@@ -51,13 +52,28 @@ type ClientGetter interface {
...
+	TableAdminClient(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+	TableClientForWatch(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+	TableAdminClientForWatch(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+}
+
+type WarningBuffer struct {
+	Warnings []types.Warning
+}
+
+func (w *WarningBuffer) HandleWarningHeader(code int, agent string, text string) {
```
[the remark this answers is not in the document]
Comment (crobby) [reply pkg/stores/proxy/proxy_store.go:68]: Yeah...just not in our code. It's what gets called when the actual client sees a warning header

File: pkg/stores/proxy/proxy_store.go [hunk 5]
```diff
@@ -51,13 +52,28 @@ type ClientGetter interface {
...
+	Client(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+	DynamicClient(ctx *types.APIRequest, warningHandler rest.WarningHandler) (dynamic.Interface, error)
+	AdminClient(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+	TableClient(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+	TableAdminClient(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+	TableClientForWatch(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+	TableAdminClientForWatch(ctx *types.APIRequest, schema *types.APISchema, namespace string, warningHandler rest.WarningHandler) (dynamic.ResourceInterface, error)
+}
+
+type WarningBuffer struct {
```
[the remark this answers is not in the document]
Comment (crobby) [reply pkg/stores/proxy/proxy_store.go]: Nope.  Updated accordingly

File: pkg/stores/proxy/proxy_store.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/stores/proxy/proxy_store.go]: Sorry, just to be sure, if it's called by someone else (who and how?), then does the type need to be exported? I am not finding any calls of this method anywhere across the org.

[the remark this answers is not in the document]
Comment (crobby) [reply pkg/stores/proxy/proxy_store.go]: When we create a client, we create a &warningBuffer to be used for that client.  That warningBuffer has the HandleWarningHeader function that gets called down in client-go at runtime.  I've witnessed this miracle in the debugger.

**Review by maxsokolovsky — APPROVED** [review]

**Pull request merged**